### PR TITLE
Adds import from JSON, common error handling, and updates Pillow

### PR DESCRIPTION
**Pull request by sntxerror**
Add import from JSON for cases, when user wants to add some tracks manually.

### CODE REVIEW

File: importer.py [hunk 1]
```diff
@@ -90,21 +91,29 @@ def __init__(self, spotify_client, yandex_client, ignore_list, strict_search):
...
+            item_name = item
+            type_ = 'track'  # Default type for string items
+            artists = []  # Default artists for string items
+        # else it is an object from Yandex
+        else:
+            type_ = item.__class__.__name__.casefold()
+            item_name = item.name if isinstance(item, Artist) else f'{", ".join([artist.name for artist in item.artists])} - {item.title}'
+            artists = item.artists  # Artists for Yandex items
+
+            if len(item_name) > 100:
```
Comment (MarshalX) [inline importer.py:107]: missed comment "#A workaround for when "? was presented before changes

File: importer.py [hunk 2]
```diff
@@ -216,6 +225,40 @@ def print_not_imported(self):
             logger.info(f'{section}:')
             for item in items:
                 logger.info(item)
+    
+    def import_from_json(self, file_path):
+        with open(file_path, "r") as file:
```
Comment (MarshalX) [inline importer.py]: pls specify encoding arg to utf-8

File: importer.py [hunk 3]
```diff
@@ -216,6 +225,40 @@ def print_not_imported(self):
...
+                logger.info('OK')
+            except NotFoundException as exception:
+                not_imported.append(exception.item_name)
+                logger.warning('NO')
+            except SpotifyException:
+                not_imported.append(query)
+                logger.warning('NO')
+
+        # Create a new playlist
+        playlist_name = "Imported from VK"
```
Comment (MarshalX) [inline importer.py]: from **VK**? perhaps should be "from JSON"

File: importer.py (hunk 2 above)
Comment (MarshalX) [reply importer.py]: mb single quotes in  'r'. much with the current code base

File: importer.py [hunk 4]
```diff
@@ -226,7 +269,7 @@ def print_not_imported(self):
     spotify_oauth.add_argument('--id', required=True, help='Client ID of your Spotify app')
     spotify_oauth.add_argument('--secret', required=True, help='Client Secret of your Spotify app')
 
-    parser.add_argument('-t', '--token', required=True, help='Token from music.yandex.com account')
```
Comment (MarshalX) [inline importer.py:229]: why not required anymore?

File: importer.py [hunk 5]
```diff
@@ -235,18 +278,27 @@ def print_not_imported(self):
...
+            client_id=arguments.id,
+            client_secret=arguments.secret,
+            redirect_uri=REDIRECT_URI,
+            scope='playlist-modify-public, user-library-modify, user-follow-modify, ugc-image-upload',
+            username=arguments.spotify,
+            cache_path='cache.txt'
+        )
+        spotify_client_ = spotipy.Spotify(auth_manager=auth_manager, requests_timeout=arguments.timeout)
+
+        yandex_client_ = Client(arguments.token)
```
Comment (MarshalX) [inline importer.py]: token looks required as before but the argument is optional now

File: importer.py [hunk 6]
```diff
@@ -235,18 +278,27 @@ def print_not_imported(self):
 
     parser.add_argument('-S', '--strict-artists-search', help='Search for an exact match of all artists', default=False)
 
+    parser.add_argument('-j', '--json', help='JSON file to import tracks from')
```
Comment (MarshalX) [inline importer.py]: mb `--json-file`/`--json-path` instead?

**Review by MarshalX — COMMENTED** [review]
good job! thank you. looks like the project uses single quotes. pls respect it

File: importer.py (hunk 4 above)
Comment (sntxerror) [reply importer.py:229]: cause this way one could import from JSON only

File: importer.py (hunk 5 above)
Comment (sntxerror) [reply importer.py]: as said before, this change allows users to use JSON only if needed

File: importer.py (hunk 1 above)
Comment (sntxerror) [reply importer.py:107]: 8bbbc51

File: importer.py (hunk 2 above)
Comment (sntxerror) [reply importer.py]: 8bbbc51

File: importer.py (hunk 6 above)
Comment (sntxerror) [reply importer.py]: 8bbbc51

File: importer.py (hunk 3 above)
Comment (sntxerror) [reply importer.py]: 8bbbc51

File: importer.py (hunk 4 above)
Comment (MarshalX) [reply importer.py:229]: understood, but why do we create yandex_music client when there is no token in args? did i miss smth? why do we need ym client for import from JSON

Comment (sntxerror) [reply importer.py:229]: c62e34c

Comment (sntxerror) [reply importer.py:229]: 9ba9417

File: importer.py [hunk 7]
```diff
@@ -235,18 +279,32 @@ def print_not_imported(self):
...
+        spotify_client_ = spotipy.Spotify(auth_manager=auth_manager, requests_timeout=arguments.timeout)
+        yandex_client_ = None
+
+        if arguments.token:
+            yandex_client_ = Client(arguments.token)
+            yandex_client_.init()
+
+        importer = Importer(spotify_client_, yandex_client_, arguments.ignore, arguments.strict_artists_search)
+
+        if arguments.json:
```
Comment (MarshalX) [inline importer.py]: `if arguments.json_path` instead?

File: importer.py [hunk 8]
```diff
@@ -235,18 +279,32 @@ def print_not_imported(self):
...
+            redirect_uri=REDIRECT_URI,
+            scope='playlist-modify-public, user-library-modify, user-follow-modify, ugc-image-upload',
+            username=arguments.spotify,
+            cache_path='cache.txt'
+        )
+
+        spotify_client_ = spotipy.Spotify(auth_manager=auth_manager, requests_timeout=arguments.timeout)
+        yandex_client_ = None
+
+        if arguments.token:
```
Comment (MarshalX) [inline importer.py:302]: if not json_path and not token:
    raise error(you must provide token for import from YM)

File: importer.py [hunk 9]
```diff
@@ -205,9 +215,9 @@ def save_artists_callback(importer, spotify_artists):
         self._add_items_to_spotify(artists, self.not_imported['Artists'], save_artists_callback)
 
     def import_all(self):
-        for item in self._importing_items.values():
-            item()
-
+        if self.yandex_client is not None:
```
Comment (MarshalX) [inline importer.py]: we should not call `import_all` if there is no YM client. the suggested if statement in the prev comment should help

Comment (sntxerror) [reply importer.py]: 3c1d76b

File: importer.py (hunk 8 above)
Comment (sntxerror) [reply importer.py:302]: 3c1d76b

File: importer.py (hunk 7 above)
Comment (sntxerror) [reply importer.py]: 3c1d76b

**Review by MarshalX — APPROVED** [review]

**Pull request merged**